### PR TITLE
Post create data set

### PR DESCRIPTION
**Pull request by jcbashdown**
Working post to create data set api end point !WITH VALIDATIONS! Oooh.

Includes refactoring of existing routes into resource view

Leaves out update for now

Does not move sub resources into resource view for now. I had refactor fatigue in my brain.

### CODE REVIEW

**Comment by jcbashdown** [comment]
Just testing this out, i'm wondering if this should create data groups by default when they don't exist?

**Comment by tombooth** [comment]
I think that might be a bit much, but it would output something so a machine can tell that is the case

**Comment by jcbashdown** [comment]
It outputs a reasonably nicely formatted error. I'll leave it at that for now.

Just fixing a bug I've just spotted.

File: stagecraft/libs/views/resource.py [hunk 1]
```diff
@@ -23,10 +23,12 @@
...
 
-def resource_url(ident, cls):
+def resource_url(ident, cls, finder_matcher=None):
+    finder_matcher = finder_matcher if finder_matcher else '<id>{}'.format(
+        UUID_RE_STRING)
     return url(
-        r'^{}(?:/(?P<id>{})(?:/(?P<sub_resource>[a-z]+))?)?'.format(
-            ident, UUID_RE_STRING),
+        r'^{}(?:/(?P{})(?:/(?P<sub_resource>[a-z]+))?)?'.format(
+            ident, finder_matcher),
```
Comment (tombooth) [inline stagecraft/libs/views/resource.py]: Could this be id_matcher to show its relation to the id of the model

File: stagecraft/apps/datasets/views/data_set.py [hunk 2]
```diff
@@ -1,44 +1,201 @@
...
+        },
+        "required": ["data_type", "data_group"],
+        "additionalProperties": False,
+    }
 
-    return HttpResponse(json_str, content_type='application/json')
+    @method_decorator(permission_required('signin'))
+    @method_decorator(never_cache)
+    @method_decorator(vary_on_headers('Authorization'))
+    def get(self, user, request, **kwargs):
```
Comment (tombooth) [inline stagecraft/apps/datasets/views/data_set.py]: Hope come you overrode get() and not by_id()?

File: stagecraft/apps/datasets/views/data_set.py [hunk 3]
```diff
@@ -1,44 +1,201 @@
...
+            return self.list(user, request)
+
+        json_str = to_json(data_set.serialize())
+
+        return HttpResponse(json_str, content_type='application/json')
+
+    # No longer called directly - happens through get which is protected.
+    def list(self, user, request):
+        # 400 if any query string keys were not in allowed set
+        if not set(request.GET).issubset(self.list_filters):
```
Comment (tombooth) [inline stagecraft/apps/datasets/views/data_set.py]: It might be good to move this into ResourceView as it strikes me as useful for other subclasses

Comment (tombooth) [inline stagecraft/apps/datasets/views/data_set.py]: Given that, is the only bit in this that is unique the `if 'admin'` bit?

**Comment by tombooth** [comment]
It would be good to move the model_json stuff from post() into update_model as in https://github.com/alphagov/stagecraft/blob/master/stagecraft/apps/organisation/views.py#L97

Would also be useful to have some more explanation about the user=None. Why does it blow up? Especially in places with an authorization decorator that should always return a user parameter

**Comment by tombooth** [comment]
The moving of the post() code might remove the model_json bits from post() in ResourceView

**Comment by tombooth** [comment]
:+1: :eggplant:

**Pull request merged**